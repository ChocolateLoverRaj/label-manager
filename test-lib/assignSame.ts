const assignSame = <T>(obj1: T, obj2: Partial<T>): T => Object.assign(obj1, obj2)

export default assignSame
