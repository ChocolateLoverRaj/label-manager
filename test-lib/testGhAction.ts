import { spawn } from 'child_process'
import streamToString from 'stream-to-string'
import { once } from 'events'
import { EOL } from 'os'
import { join } from 'path'
import { writeFile } from 'jsonfile'
import defaults from 'object.defaults'
import getPort from 'get-port'
import express from 'express'
import { createServer } from 'http'

interface Result {
  outputs: Record<string, string>
  stdout: string
}

export interface Commit {
  message: string
}

export interface Label {
  name: string
}

export interface PullRequest {
  commits: Commit[]
  labels: Label[]
}

interface Repo {
  pullRequests: Record<number, PullRequest>
  token?: string
}

interface PrEvent {
  number: number
}

interface PushEvent {
  commits: Commit[]
}

interface Options {
  inputs: Record<string, string>
  event: PrEvent | PushEvent
  repo: Repo
  env: Record<string, string>
}

const testGhAction = async (file: string, partialOptions: Partial<Options> = {}): Promise<Result> => {
  const options: Options = defaults(partialOptions, { inputs: [], event: { number: 1 }, repo: { pullRequests: {} } })
  const { event, inputs, repo } = options
  const eventFile = join(__dirname, '../test-tmp/event.json')
  const port = await getPort()

  const server = express()
  const fakeGhServer = createServer(server).listen(port)

  server.get('/repos/ChocolateLoverRaj/test-semver/pulls/:number/commits', (req, res) => {
    const n = parseInt(req.params.number)
    const pr = repo.pullRequests[n]
    if (pr === undefined) {
      res.sendStatus(404)
      return
    }
    res.json(pr.commits.map(commit => ({ commit })))
  })

  server.post('/repos/ChocolateLoverRaj/test-semver/issues/:number/labels', express.json(), (req, res) => {
    const pr = repo.pullRequests[parseInt(req.params.number)]
    if (pr === undefined) return res.sendStatus(404)
    pr.labels.push(...(req.body.labels as string[]).map(name => ({ name })))
    res.end()
  })

  server.delete('/repos/ChocolateLoverRaj/test-semver/issues/:number/labels/:name', express.json(), (req, res) => {
    const pr = repo.pullRequests[parseInt(req.params.number)]
    if (pr === undefined) return res.sendStatus(404)
    const labelIndex = pr.labels.findIndex(({ name }) => name === req.params.name)
    if (labelIndex === -1) return res.sendStatus(404)
    pr.labels.splice(labelIndex, 1)
    res.end()
  })

  await Promise.all([
    writeFile(eventFile, {
      pull_request: 'number' in event
        ? {
            commits: repo.pullRequests[event.number].commits.length,
            number: event.number,
            labels: repo.pullRequests[event.number].labels
          }
        : undefined,
      commits: 'commits' in event ? event.commits : undefined,
      repository: {
        name: 'test-semver',
        owner: {
          login: 'ChocolateLoverRaj'
        }
      }
    }, { spaces: 2 }),
    once(fakeGhServer, 'listening')
  ])

  const c = spawn('node', [file], {
    env: {
      ...process.env,
      GITHUB_EVENT_PATH: eventFile,
      GITHUB_API_URL: `http://localhost:${port}`,
      ...Object.fromEntries(Object.entries(inputs).map(([k, v]) => [`INPUT_${k.toUpperCase()}`, v])),
      ...options.env
    }
  })
  const stdoutStrPromise = streamToString(c.stdout)
  const stderrStr = streamToString(c.stderr)
  const [code] = await once(c, 'exit')
  fakeGhServer.close()
  if (code !== 0) {
    throw new Error(
      `Process exited with code ${code as number}\n` +
      'Stderr is below:\n' +
      await stderrStr
    )
  }
  const setOutputStr = '::set-output name='
  const stdoutStr = (await stdoutStrPromise)
  return {
    outputs: Object.fromEntries(stdoutStr
      .split(EOL)
      .map(str => {
        if (str.startsWith(setOutputStr)) {
          const keyValue = str.slice(setOutputStr.length).split('::')
          if (keyValue.length !== 2) throw new Error('Bad set output syntax')
          return keyValue
        }
        return undefined
      })
      .filter(v => v !== undefined) as any
    ),
    stdout: stdoutStr
  }
}

export default testGhAction
