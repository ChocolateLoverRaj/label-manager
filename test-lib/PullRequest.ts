import { Commit, Label, PullRequest as IPullRequest } from './testGhAction'

class PullRequest implements IPullRequest {
  commits: Commit[] = []
  labels: Label[] = []
}

export default PullRequest
