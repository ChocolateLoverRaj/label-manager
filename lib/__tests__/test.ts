import testGhAction from '../../test-lib/testGhAction'
import { join } from 'path'
import PullRequest from '../../test-lib/PullRequest'
import assignSame from '../../test-lib/assignSame'

const mainFilePath = join(__dirname, '../../dist/index.js')

test('pull request', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: assignSame(new PullRequest(), {
          commits: [{
            message: 'Feature: add triangles'
          }, {
            message: 'Breaking: replaced triangle with polygon'
          }, {
            message: 'Fix: polygons have minimum of 3 sides'
          }, {
            message: 'Chore: updated GitHub action script'
          }]
        })
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('major')
})

test('push', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      commits: [{
        message: 'Feat: add images'
      }, {
        message: 'Fix: fix bug about images'
      }]
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('minor')
})

test('invalid commit message', async () => {
  await expect(testGhAction(join(__dirname, '../../dist/index.js'), {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: assignSame(new PullRequest(), {
          commits: [{ message: 'made it better' }]
        })
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })).rejects.toMatchSnapshot()
})
