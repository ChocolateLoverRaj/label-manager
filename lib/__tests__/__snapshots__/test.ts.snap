// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid commit message 1`] = `
[Error: Process exited with code 1
Stderr is below:
Invalid Commit Message: made it better
Messages should follow: <type>[optional scope]: <description>
]
`;

exports[`pull request 1`] = `
"Fetching 4 commits.
Parsing commit messages
Message: Feature: add triangles. Increment: minor.
Message: Breaking: replaced triangle with polygon. Increment: major.
Message: Fix: polygons have minimum of 3 sides. Increment: patch.
Message: Chore: updated GitHub action script. Increment: none.
Largest increment: major

::set-output name=increment::major
"
`;

exports[`push 1`] = `
"Parsing commit messages
Message: Feat: add images. Increment: minor.
Message: Fix: fix bug about images. Increment: patch.
Largest increment: minor

::set-output name=increment::minor
"
`;
